lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,483 @@
+        dissimilarities: Nx.tensor([1.0, 2.0, 2.0, 4.0]),
+        num_points: 5,
+        sizes: Nx.tensor([2, 2, 3, 5])
+      }
+  """
+  deftransform fit(%Nx.Tensor{} = data, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @fit_opts_schema)
+    dissimilarity = opts[:dissimilarity]
+    linkage = opts[:linkage]
+
+    dissimilarity_fun =
+      case dissimilarity do
+        # :precomputed -> &Function.identity/1
+        :euclidean -> &pairwise_euclidean/1
+      end
+
+    update_fun =
+      case linkage do
+        :average -> &average/6
+        # :centroid -> &centroid/6
+        :complete -> &complete/6
+        # :median -> &median/6
+        :single -> &single/6
+        :ward -> &ward/6
+        :weighted -> &weighted/6
+      end
+
+    dendrogram_fun =
+      case linkage do
+        # TODO: :centroid, :median
+        l when l in [:average, :complete, :single, :ward, :weighted] ->
+          &parallel_nearest_neighbor/3
+      end
+
+    n =
+      case Nx.shape(data) do
+        {n, _num_features} ->
+          n
+
+        other ->
+          raise ArgumentError,
+                "Expected a rank 2 (`{num_obs, num_features}`) tensor, found shape: #{inspect(other)}."
+      end
+
+    if n < 3 do
+      raise ArgumentError, "Must have a minimum of 3 data points, found: #{n}."
+    end
+
+    {clades, diss, sizes} = dendrogram_fun.(data, dissimilarity_fun, update_fun)
+
+    %__MODULE__{
+      clades: clades,
+      dissimilarities: diss,
+      num_points: n,
+      sizes: sizes
+    }
+  end
+
+  # Clade functions
+
+  defnp parallel_nearest_neighbor(data, dissimilarity_fun, update_fun) do
+    pairwise = dissimilarity_fun.(data)
+    {n, _} = Nx.shape(pairwise)
+    pairwise = Nx.broadcast(:infinity, {n}) |> Nx.make_diagonal() |> Nx.add(pairwise)
+    clades = Nx.broadcast(-1, {n - 1, 2})
+    sizes = Nx.broadcast(1, {2 * n - 1})
+    pointers = Nx.broadcast(-1, {2 * n - 2})
+    diss = Nx.tensor(:infinity, type: Nx.type(pairwise)) |> Nx.broadcast({n - 1})
+
+    {{clades, diss, sizes}, _} =
+      while {{clades, diss, sizes}, {count = 0, pointers, pairwise}}, count < n - 1 do
+        # Indexes of who I am nearest to
+        nearest = Nx.argmin(pairwise, axis: 1)
+
+        # Take who I am nearest to is nearest to
+        nearest_of_nearest = Nx.take(nearest, nearest)
+
+        # If the entry is pointing back at me, then we are a clade
+        clades_selector = Nx.equal(nearest_of_nearest, Nx.iota({n}))
+
+        # Now let's get the links that form clades.
+        # They are bidirectional but let's keep only one side.
+        links = Nx.select(clades_selector and nearest > nearest_of_nearest, nearest, n)
+
+        {clades, count, pointers, pairwise, diss, sizes} =
+          merge_clades(clades, count, pointers, pairwise, diss, sizes, links, n, update_fun)
+
+        {{clades, diss, sizes}, {count, pointers, pairwise}}
+      end
+
+    sizes = sizes[n..(2 * n - 2)]
+    perm = Nx.argsort(diss, type: :u32)
```

Comment:
If you need this to be stable, please pass `stable: true`, as the default will change on Nx v0.7 to mirror Numpy. Perhaps best to pass it as true or false anyway to keep it explicit.

Reply:
Noted! I don't actually think it needs to be stable, but I'll make it explicit.

Reply:
[cb15a91](https://github.com/elixir-nx/scholar/pull/187/commits/cb15a91423ac6560fb8aed87ee320267a070234f)